consensus/hotstuff/safetyrules/safety_rules_test.go

```diff
@@ -1,165 +1,526 @@
+	timeout, err := s.safety.ProduceTimeout(view, newestQC, nil)
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), timeout)
+
+	// voting in same view after producing timeout is not allowed
+	vote, err := s.safety.ProduceVote(s.proposal, view)
+	require.True(s.T(), model.IsNoVoteError(err))
+	require.Nil(s.T(), vote)
+
+	s.signer.AssertExpectations(s.T())
+	s.persister.AssertExpectations(s.T())
+}
+
+// TestProduceTimeout_ShouldTimeout tests that we can produce timeout in cases where
+// last view was successful or not. Also tests last timeout caching.
+func (s *SafetyRulesTestSuite) TestProduceTimeout_ShouldTimeout() {
+	view := s.proposal.Block.View
+	newestQC := helper.MakeQC(helper.WithQCView(view - 1))
+	expectedTimeout := &model.TimeoutObject{
+		View:     view,
+		NewestQC: newestQC,
+	}
+
+	expectedSafetyData := &hotstuff.SafetyData{
+		LockedOneChainView:      s.safetyData.LockedOneChainView,
+		HighestAcknowledgedView: view,
+		LastTimeout:             expectedTimeout,
+	}
+	s.signer.On("CreateTimeout", view, newestQC, (*flow.TimeoutCertificate)(nil)).Return(expectedTimeout, nil).Once()
+	s.persister.On("PutSafetyData", expectedSafetyData).Return(nil).Once()
+	timeout, err := s.safety.ProduceTimeout(view, newestQC, nil)
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), expectedTimeout, timeout)
+
+	s.persister.AssertCalled(s.T(), "PutSafetyData", expectedSafetyData)
+
+	// producing timeout with same arguments should return cached version
+	otherTimeout, err := s.safety.ProduceTimeout(view, newestQC, nil)
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), timeout, otherTimeout)
+
+	// to create new TO we need to provide a TC
+	lastViewTC := helper.MakeTC(helper.WithTCView(view),
+		helper.WithTCHighestQC(newestQC))
+
+	expectedTimeout = &model.TimeoutObject{
+		View:       view + 1,
+		NewestQC:   newestQC,
+		LastViewTC: lastViewTC,
+	}
+	s.signer.On("CreateTimeout", view+1, newestQC, lastViewTC).Return(expectedTimeout, nil).Once()
+	expectedSafetyData = &hotstuff.SafetyData{
+		LockedOneChainView:      s.safetyData.LockedOneChainView,
+		HighestAcknowledgedView: view + 1,
+		LastTimeout:             expectedTimeout,
+	}
+	s.persister.On("PutSafetyData", expectedSafetyData).Return(nil).Once()
+
+	// creating new timeout should invalidate cache
+	otherTimeout, err = s.safety.ProduceTimeout(view+1, newestQC, lastViewTC)
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), otherTimeout)
+
+	// creating timeout for previous view(that was already cached) should result in error
+	timeout, err = s.safety.ProduceTimeout(view, newestQC, nil)
+	require.Error(s.T(), err)
+	require.Nil(s.T(), timeout)
 }
 
-func testVotingAgain(t *testing.T) {
-	blockView, curView, lastVotedView, isBlockSafe, isCommitteeMember := uint64(3), uint64(3), uint64(2), true, true
+// TestProduceTimeout_NotSafeToTimeout tests that we don't produce a timeout when it's not safe
+func (s *SafetyRulesTestSuite) TestProduceTimeout_NotSafeToTimeout() {
```

Comment:
I think we expect all these cases to lead to exceptions (?)

```suggestion
// We expect that the EventHandler to feed only request timeouts for the current view, providing valid set of inputs.
// Hence, the cases tested here would be symptoms of an internal bugs, and therefore should not result in an NoVoteError.
func (s *SafetyRulesTestSuite) TestProduceTimeout_NotSafeToTimeout() {
```
Furthermore, I think we should add the assertion everywhere where we have `require.Error(s.T(), err)` in this test:
```golang
require.False(s.T(), model.IsNoVoteError(err))
```